src/plugins/gds_mt/gds_mt_backend.cpp

```diff
@@ -0,0 +1,296 @@
+            if (!base_addr) {
+                return NIXL_ERR_INVALID_PARAM;
+            }
+            total_size = remote[i].len;
+            base_offset = (size_t)local[i].addr;
+
+            auto it = gds_mt_file_map_.find(local[i].devId);
+            if (it == gds_mt_file_map_.end()) {
+                NIXL_ERROR << "GDS_MT: error: file handle not found";
+                return NIXL_ERR_NOT_FOUND;
+            }
+            cu_fhandle = it->second->cu_fhandle;
+        } else {
+            base_addr = (void*)local[i].addr;
+            if (!base_addr) {
+                return NIXL_ERR_INVALID_PARAM;
+            }
+            total_size = local[i].len;
+            base_offset = (size_t)remote[i].addr;
+
+            auto it = gds_mt_file_map_.find(remote[i].devId);
+            if (it == gds_mt_file_map_.end()) {
+                NIXL_ERROR << "GDS_MT: error: file handle not found";
+                return NIXL_ERR_NOT_FOUND;
+            }
+            cu_fhandle = it->second->cu_fhandle;
+        }
+
+        gds_mt_handle->request_list.emplace_back(base_addr, total_size, base_offset, cu_fhandle,
+            (operation == NIXL_READ) ? CUFILE_READ : CUFILE_WRITE);
+    }
+
+    if (gds_mt_handle->request_list.empty()) {
+        return NIXL_ERR_INVALID_PARAM;
+    }
+    for (GdsMtTransferRequestH& req : gds_mt_handle->request_list) {
+        GdsMtTransferRequestH* captured_req = &req;
+        gds_mt_handle->taskflow.emplace([captured_req]() {
+            runCuFileOp(captured_req);
+        });
+    }
+
+
+    handle = gds_mt_handle.release();
+    return NIXL_SUCCESS;
+}
+
+nixl_status_t nixlGdsMtEngine::postXfer(const nixl_xfer_op_t &operation,
+                                        const nixl_meta_dlist_t &local,
+                                        const nixl_meta_dlist_t &remote,
+                                        const std::string &remote_agent,
+                                        nixlBackendReqH* &handle,
+                                        const nixl_opt_b_args_t* opt_args) const
+{
+    nixlGdsMtBackendReqH* gds_mt_handle = (nixlGdsMtBackendReqH*)handle;
+    if (!gds_mt_handle) {
+        NIXL_ERROR << "GDS_MT: error: invalid handle";
+        return NIXL_ERR_INVALID_PARAM;
+    }
+    if (gds_mt_handle->request_list.empty()) {
+        NIXL_ERROR << "GDS_MT: error: empty request list for Xfer";
+        return NIXL_ERR_INVALID_PARAM;
+    }
+    gds_mt_handle->running_transfer = executor_->run(gds_mt_handle->taskflow);
+    return NIXL_IN_PROG;
+}
+
+nixl_status_t nixlGdsMtEngine::checkXfer(nixlBackendReqH* handle) const
+{
+    nixlGdsMtBackendReqH *gds_mt_handle = (nixlGdsMtBackendReqH *)handle;
+    if (gds_mt_handle->request_list.empty()) {
+        return NIXL_SUCCESS;
+    }
+    if (gds_mt_handle->running_transfer.wait_for(std::chrono::seconds(0)) != std::future_status::ready) {
+        return NIXL_IN_PROG;
+    }
+    gds_mt_handle->running_transfer.get();
```

Comment:
AFAIU this "waits" for future to Finish, will that end up blocking in checkXfer in any scenario?